### PR TITLE
Nixos integration test ports

### PR DESCRIPTION
**Pull request by tfc**
##### Motivation for this change

Processing the task list in issue #72828.

###### Things done

- implemented `require_active_unit`, `get_window_names`, and `wait_for_window`.
- ported the tests `matrix-synapse.nix` and `firefox.nix`

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli @marijanp

### CODE REVIEW

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -642,6 +655,27 @@ def wait_for_x(self):
                 if status == 0:
                     return
 
+    def get_window_names(self):
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: This probably is internal API, so `_get_window_names`?

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -642,6 +655,27 @@ def wait_for_x(self):
...
+            if last_try:
+                self.log(
+                    "Last chance to match {} on the window list,".format(regexp)
+                    + " which currently contains: "
+                    + ", ".join(names)
+                )
+            return any(pattern.search(name) for name in names)
+
+        with self.nested("Waiting for a window to appear"):
+            retry(window_is_visible)
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: Can we inline the window_is_visible function in a lambda here maybe?
Is that log line present in the perl code too?

File: nixos/lib/test-driver/test-driver.py [hunk 3]
```diff
@@ -344,6 +345,18 @@ def systemctl(self, q, user=None):
             )
         return self.execute("systemctl {}".format(q))
 
+    def require_active_unit(self, unit):
```
Comment (flokli) [inline nixos/lib/test-driver/test-driver.py]: We could probably rename this to `require_unit_state`, with an optional parameter defaulting to `active`.

File: nixos/lib/test-driver/test-driver.py (hunk 2 above)
Comment (tfc) [reply nixos/lib/test-driver/test-driver.py]: i had the impression that multi-line-lambdas are considered bad style in python. the log line is present in the perl code, yes.

File: nixos/lib/test-driver/test-driver.py [hunk 4]
```diff
@@ -655,6 +655,27 @@ def wait_for_x(self):
                 if status == 0:
                     return
 
+    def get_window_names(self):
+        return self.succeed(
+            r"xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'"
```
Comment (worldofpeace) [inline nixos/lib/test-driver/test-driver.py]: can we not use `xwininfo`? I'm sure there's a python library just for this.

**Comment by worldofpeace** [comment]
With https://github.com/NixOS/nixpkgs/pull/72835/commits/45b339b736380fe007721a460b2d2ac9791d8f5c I can port all the DE tests and installed tests for gnome to the python driver.

File: nixos/lib/test-driver/test-driver.py (hunk 4 above)
Comment (tfc) [reply nixos/lib/test-driver/test-driver.py]: hey @worldofpeace, thank you for the suggestion. It surely makes sense to use better alternatives if there are any, but questioning the methods of the perl scripts, researching what other libraries are there and doing things differently for all the 300 tests that have to be ported significantly blows up the amount of work we have to do here.

Can we please discuss such things offline and find consensus on describing in follow-up tickets on how to improve such things once the port is complete - or find people who like to help out on such things?

**Comment by tfc** [comment]
> With [45b339b](https://github.com/NixOS/nixpkgs/commit/45b339b736380fe007721a460b2d2ac9791d8f5c) I can port all the DE tests and installed tests for gnome to the python driver.
> (unless you've done this locally already 😄 )

Not yet, but it would make sense to ask in #nixos-python-test-driver on freenode if someone is on test *X* already... had some clashes in the past days. :-D

**Comment by tfc** [comment]
@domenkozar

**Review by domenkozar — APPROVED** [review]

**Review by worldofpeace — APPROVED** [review]

File: nixos/lib/test-driver/test-driver.py [hunk 5]
```diff
@@ -344,6 +345,18 @@ def systemctl(self, q, user=None):
             )
         return self.execute("systemctl {}".format(q))
 
+    def require_unit_state(self, unit, require_state="active"):
+        with self.nested(
+            "checking if unit ‘{}’ has reached state '{}'".format(unit, require_state)
+        ):
+            info = self.get_unit_info(unit)
```
Comment (worldofpeace) [inline nixos/lib/test-driver/test-driver.py]: Should this be allowed to accept `user` also? (like wait_for_unit)

**Pull request merged**